force-app/main/default/classes/Circle.cls

```diff
@@ -0,0 +1,55 @@
+/**
+* @description encapsulates a circle, which is immutable. 
+* It is be possible to construct a circle by providing either it’s radius 
+* or a string representation of its diameter
+*/
+public with sharing class Circle 
+{
+    private Static final Decimal pi = 3.14159; // PI to 5dp
+    public final Decimal radius;
+    public final Decimal circumference;
+    public final Double surfaceArea;
+
+    public Circle(Decimal radius) 
+    {
+        if (radius > 0)
+        {
+            This.radius = radius;
+            This.circumference = This.calculateCircumference();
+            This.surfaceArea = This.calculateSurfaceArea();
+        }
+        else
+        {
+            Throw new CirclePositiveRadiusException('The radius must be > 0');
+        }
+    }
+
+    public Circle(String diameter) 
+    {   
+        if (0.5 * decimal.valueOf(diameter)> 0)
+        {
+            This.radius = 0.5 * decimal.valueOf(diameter);
+            This.circumference = This.calculateCircumference();
+            This.surfaceArea = This.calculateSurfaceArea();
+        }
+        else
+        {
+            Throw new CirclePositiveRadiusException('The diameter must be > 0');
```

Comment:
Here's a good example of why exceptions that are too granular aren't great - you are throwing a radius exception because of a problem with the diameter. While they are related, in practice you'd probably need a diameter specific exception instead if you continued this route.